### PR TITLE
Add composite filters in support of OR queries

### PR DESCRIPTION
**Pull request by MarkDuckworth**
Add composite filters. Port of https://github.com/firebase/firebase-android-sdk/pull/3290

### CODE REVIEW

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [inline packages/firestore/src/lite-api/query.ts:993]: As you mentioned last time, it's not great that we need to do this validation, and points to room for improvement in the API.

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [inline packages/firestore/src/lite-api/query.ts:281]: did you figure out the deal with `and` vs `andQuery` ?

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [inline packages/firestore/src/lite-api/query.ts:206]: can we use `CompositeOperator` ?

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [inline packages/firestore/src/lite-api/query.ts:258]: is there a way to _**not**_ expose `and()` and `or()` publicly yet (and still be able to invoke them in the tests) ?

Android makes that possible using `@RestrictTo(RestrictTo.Scope.LIBRARY)`. But not sure if JS has a way to do this.

**Review by ehsannas — CHANGES_REQUESTED** [review]
This is a great start, and generally looks good to me. Some nits/comments outlined below.

File: packages/firestore/src/core/target.ts
Comment (MarkDuckworth) [reply packages/firestore/src/core/target.ts:742]: @ehsannas, I just looked at the implementation for mutability in the Android SDK. Are you not concerned that the `filters` collection is mutable?

File: packages/firestore/src/core/target.ts
Comment (MarkDuckworth) [reply packages/firestore/src/core/target.ts:754]: done

File: packages/firestore/src/core/target.ts
Comment (MarkDuckworth) [reply packages/firestore/src/core/target.ts:814]: Good call on using `${filter.op}` directly.

IMHO, it's better to compute the `cannoicalIdsString` in the previous line. That kind of coding improves debuggability. But I'll use your full suggestion if you feel strongly about it.

File: packages/firestore/src/core/target.ts
Comment (MarkDuckworth) [reply packages/firestore/src/core/target.ts:540]: Yes! This would improved the readability/navigation of the code. However, I want to do this refactoring in another PR with no other changes. I'm adding a TODO for this refactoring.

File: packages/firestore/src/lite-api/query.ts
Comment (MarkDuckworth) [reply packages/firestore/src/lite-api/query.ts:993]: I'm going to leave this in until we get approval on a different API that doesn't need this. However, I realized that I need to test this validation.

File: packages/firestore/src/lite-api/query.ts
Comment (MarkDuckworth) [reply packages/firestore/src/lite-api/query.ts:281]: I was either using the tools incorrectly or missing an export somewhere. But I'm now able to use `and` as the design specified.

File: packages/firestore/src/lite-api/query.ts
Comment (MarkDuckworth) [reply packages/firestore/src/lite-api/query.ts:206]: In QueryConstraint, `type` is defined as `abstract readonly type: QueryConstraintType;`.

It looks like I could use `CompositeOperator`, but I don't think mixing the types there is a best practice. Further, the current code is consistent with other QueryConstraint subtypes, such as [QueryLimitConstraint](https://github.com/firebase/firebase-js-sdk/blob/7cad614ec2d2a34b40a3c24443c4f35571e3e68c/packages/firestore/src/lite-api/query.ts#L283).

File: packages/firestore/src/lite-api/query.ts
Comment (MarkDuckworth) [reply packages/firestore/src/lite-api/query.ts:258]: My understanding is by keeping it out of [api.ts](https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/api.ts), it's still internal, but can be used for testing. Please correct me if you see a gap in my understanding.

File: packages/firestore/src/core/target.ts
Comment (ehsannas) [reply packages/firestore/src/core/target.ts:540]: 👍

File: packages/firestore/src/core/target.ts
Comment (ehsannas) [reply packages/firestore/src/core/target.ts:742]: We don't do any mutations on these filters, but yeah using an `ImmutableList<Filter>` would have been better. I'll post a PR for that.

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [reply packages/firestore/src/lite-api/query.ts:206]: Ah, yes. Good call.

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [reply packages/firestore/src/lite-api/query.ts:258]: Yep; just looked into it a bit more. (as a side, we also have ["lite" APIs](https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/lite/index.ts) for the ["Lite Web SDK"](https://firebase.google.com/docs/firestore/solutions/firestore-lite)).

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [reply packages/firestore/src/lite-api/query.ts:281]: so I'm guessing you'll replace `andQuery` with `and` here?

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [reply packages/firestore/src/lite-api/query.ts:993]: 👍

File: packages/firestore/src/core/target.ts
Comment (MarkDuckworth) [reply packages/firestore/src/core/target.ts:742]: Cool. I'm changing the filters collection to a `ReadonlyArray<Filter>` here in the Web SDK.

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [reply packages/firestore/src/lite-api/query.ts:993]: Given that `or` and `and` are currently not public, and that we expect this validateQueryFilterConstraint function to hopefully go away with the new API design, I think you can skip adding tests for it right now.

File: packages/firestore/src/core/target.ts
Comment (ehsannas) [reply packages/firestore/src/core/target.ts]: Assigned a task to you for this.

**Review by ehsannas — APPROVED** [review]

File: common/api-review/firestore-lite.api.md
Comment (ehsannas) [inline common/api-review/firestore-lite.api.md:233]: given that we are not exposing 'and' and 'or' publicly, we should remove this update to the .md file

File: common/api-review/firestore.api.md
Comment (ehsannas) [inline common/api-review/firestore.api.md:362]: given that we are not exposing 'and' and 'or' publicly, we should remove this update to the .md file

**Review by ehsannas — COMMENTED** [review]
Sorry I missed this in my first review. Given that we do want to use 'and' and 'or', but we don't want to expose them publicly, we shouldn't be updating the .md files.

File: common/api-review/firestore-lite.api.md
Comment (ehsannas) [reply common/api-review/firestore-lite.api.md:233]: oops.. I thought if something is not exported in [api.ts](https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/api.ts) then it's not considered public API.

File: packages/firestore/src/lite-api/query.ts
Comment (ehsannas) [reply packages/firestore/src/lite-api/query.ts:258]: @MarkDuckworth , IIUC based on @hsubox76 's comment, this comment still stands

File: common/api-review/firestore-lite.api.md
Comment (ehsannas) [reply common/api-review/firestore-lite.api.md:233]: Ah! Got it. Thanks

Comment (MarkDuckworth) [reply common/api-review/firestore-lite.api.md:233]: Thanks for the feedback, but I don't think there's any value in trying to address this. Although this addition of 'and' and 'or' in `QueryConstraintType` will be public, it does not make this feature usable. In fact, I'm not sure why `QueryConstraintType` is publicly exposed at all, but given that it is, I'm not going to modify that.

So, @hsubox76, @ehsannas, I'm suggesting to move forward with this PR as is.

File: common/api-review/firestore.api.md
Comment (MarkDuckworth) [reply common/api-review/firestore.api.md:362]: Merging to a feature branch so that we can avoid exposing any public API changes.

File: common/api-review/firestore-lite.api.md
Comment (MarkDuckworth) [reply common/api-review/firestore-lite.api.md:233]: Merging to a feature branch so we can avoid any public API changes until the feature is released.

**Pull request merged**
